### PR TITLE
MINIFICPP-1319 - Stream refactor

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/test/unit/FileStreamTests.cpp
Comment (adamdebreceni) [reply libminifi/test/unit/FileStreamTests.cpp:238]: done

File: libminifi/src/io/OutputStream.cpp
Comment (adamdebreceni) [reply libminifi/src/io/OutputStream.cpp]: didn't find the warnings, but added explicit cast

File: libminifi/src/io/InputStream.cpp
Comment (adamdebreceni) [reply libminifi/src/io/InputStream.cpp]: done, (and put back the previous template for reading/writing unsigned integrals)

File: libminifi/src/io/BufferStream.cpp
Comment (adamdebreceni) [reply libminifi/src/io/BufferStream.cpp:19]: done

File: libminifi/include/io/OutputStream.h
Comment (adamdebreceni) [reply libminifi/include/io/OutputStream.h]: done

Comment (szaszm) [reply libminifi/include/io/OutputStream.h]: I'd prefer a pure virtual function so that an OutputStream must be writable.

File: extensions/coap/tests/CoapC2VerifyHeartbeat.cpp
Comment (szaszm) [reply extensions/coap/tests/CoapC2VerifyHeartbeat.cpp]: Our `utils::make_unique` doesn't support arrays yet.

Comment (adamdebreceni) [reply extensions/coap/tests/CoapC2VerifyHeartbeat.cpp]: yep it does not, but `unique_ptr` and move works, just taking an array as `std::unique_ptr<uint8_t>` is... not great, so I will look into it

File: libminifi/include/io/OutputStream.h
Comment (adamdebreceni) [reply libminifi/include/io/OutputStream.h]: made it pure virtual (InputStream::read as well)

File: libminifi/include/io/Stream.h
Comment (adamdebreceni) [reply libminifi/include/io/Stream.h:46]: I'm not entirely sure where we should put them, for example `FileOutputCallback::process` takes a `BaseStream` and then calls `getBuffer` (which only makes sense for a `BufferStream`), if we would like to change the method, that it can only take a `BufferStream`, then we are up for a (not necessary infeasible) ride, I moved these methods back and forth between this Stream class and the BaseStream, `ZlibCompressStream` is only an `OutputStream` but it does implement a non-trivial `close` method, that's why they settled in `Stream`, also in the future `content_repo->write(claim)` should only return an `OutputStream`, not a `BaseStream`, and as it stands we do call `seek` on the result

I think we should investigate further with your proposed "granularization" in a separate PR

File: libminifi/include/io/ZlibStream.h
Comment (adamdebreceni) [reply libminifi/include/io/ZlibStream.h]: did it for both the ZlibStream and CRCStream

File: libminifi/include/sitetosite/Peer.h
Comment (szaszm) [inline libminifi/include/sitetosite/Peer.h:313]:
```suggestion
    stream_ = utils::exchange(other.stream_, nullptr);
```

File: libminifi/src/io/InputStream.cpp
Comment (szaszm) [inline libminifi/src/io/InputStream.cpp:50]: I know it's not new but I don't understand the purpose of `!widen`. If there is none, can we get rid of it?

File: libminifi/test/unit/Site2SiteTests.cpp
Comment (szaszm) [inline libminifi/test/unit/Site2SiteTests.cpp:97]: This looks like a fixed memory leak in the test. :+1:

File: libminifi/test/unit/SiteToSiteHelper.h
Comment (szaszm) [inline libminifi/test/unit/SiteToSiteHelper.h:40]: Avoid C-style casts. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named https://google.github.io/styleguide/cppguide.html#Casting

File: libminifi/src/sitetosite/SiteToSiteClient.cpp
Comment (adamdebreceni) [reply libminifi/src/sitetosite/SiteToSiteClient.cpp:502]: this was possibly caused by an unsuccessful rebase

File: libminifi/include/io/CRCStream.h
Comment (adamdebreceni) [reply libminifi/include/io/CRCStream.h:78]: I am not sure what is the standard-compliant behavior here, but gcc and clang both accept the following, but msvc complained for the undefined reference `A::A()`

```
struct A {
  A();  // is a declaration enough?
  A(int x) : x(x) {}

 private:
  int x;
};

struct B : virtual A {};

struct C : virtual A {};

struct D : B, C {
  D(int x): A(x) {} 
};
```

File: libminifi/include/io/OutputStream.h
Comment (adamdebreceni) [reply libminifi/include/io/OutputStream.h]: done

File: libminifi/src/io/BufferStream.cpp
Comment (adamdebreceni) [reply libminifi/src/io/BufferStream.cpp]: switched to `gsl_Expects` everywhere

File: libminifi/include/sitetosite/Peer.h
Comment (adamdebreceni) [reply libminifi/include/sitetosite/Peer.h]: done

File: libminifi/include/io/CRCStream.h
Comment (adamdebreceni) [reply libminifi/include/io/CRCStream.h]: done

File: libminifi/include/io/CRCStream.h
Comment (adamdebreceni) [reply libminifi/include/io/CRCStream.h]: done

File: libminifi/test/unit/FileStreamTests.cpp
Comment (adamdebreceni) [reply libminifi/test/unit/FileStreamTests.cpp]: done, I don't remember why the cast was needed at all

File: libminifi/test/unit/FileStreamTests.cpp
Comment (adamdebreceni) [reply libminifi/test/unit/FileStreamTests.cpp]: done

File: libminifi/test/unit/SiteToSiteHelper.h
Comment (adamdebreceni) [reply libminifi/test/unit/SiteToSiteHelper.h]: done

File: libminifi/include/io/CRCStream.h
Comment (adamdebreceni) [reply libminifi/include/io/CRCStream.h]: in light of recent changes (i.e. making the field nullable) this whole questionable stuff is unnecessary, and has been removed

File: libminifi/src/io/InputStream.cpp
Comment (adamdebreceni) [reply libminifi/src/io/InputStream.cpp:50]: it seems like there is actual need for `widen` as some parts call it with `true` while others with `false`, further investigation is necessary but I wouldn't change this in this PR

**Review by szaszm — APPROVED** [review]

**Comment by szaszm** [comment]
I'm waiting for more reviews before merging.

**Review by arpadboda — APPROVED** [review]

**Pull request closed without merging**
